stan/math/rev/core/precomputed_gradients.hpp

```diff
@@ -17,51 +17,118 @@ namespace math {
  *
  * Stan users should use function precomputed_gradients()
  * directly.
+ *
+ * @tparam ContainerOperands tuple of any container operands (var_value
+ * containing Eigen types)
+ * @tparam ContainerGradients tupleof any container gradients (Eigen types)
  */
-class precomputed_gradients_vari : public vari {
+template <typename ContainerOperands = std::tuple<>,
+          typename ContainerGradients = std::tuple<>>
+class precomputed_gradients_vari_template : public vari {
  protected:
   const size_t size_;
   vari** varis_;
   double* gradients_;
+  static_assert(std::tuple_size<ContainerOperands>::value
+                    == std::tuple_size<ContainerGradients>::value,
+                "precomputed_gradients_vari: ContainerOperands and "
+                "ContainerGradients should have same size!");
+  static constexpr size_t N_containers
+      = std::tuple_size<ContainerOperands>::value;
+  ContainerOperands container_operands_;
```

Comment:
If I'm reading [this](https://github.com/stan-dev/math/blob/5060b683bc031f102ae8ade4a451013a5e567f59/stan/math/rev/functor/operands_and_partials.hpp#L221) correctly, then I think this container will end up containing Eigen dynamic types, which is a memory leak (the destructor never gets called on varis).

Reply:
Right, I need use maps for these.

Reply:
This turned out to be a bit complicated to fix. Using pointers and maps for elements of these tuples would be a bit ugly. Instead I decided to add separate types for eigen matrices and std vectors that have dynamically allocated memory on AD stack, so they are trivially destructible. Std vectors required just required custom allocator. For Eigen matrix I subclassed `Eigen::Map` and added memory management.

I believe these will also be useful in other places.